File: PHP/111. Minimum Depth of Binary Tree.php
<?php

/**
 * @noinspection PhpIllegalPsrClassPathInspection
 * @noinspection PhpMultipleClassDeclarationsInspection
 * @noinspection PhpUnused
 */
class Solution
{
    public function minDepth(?TreeNode $root): int
    {
        if ($root === null) {
            return 0;
        }

        $depth = 1;

        $queue = new SplQueue;
        $queue->enqueue($root);

        while (! $queue->isEmpty()) {
            $size = $queue->count();

            for ($i = 0; $i < $size; $i++) {
                $node = $queue->dequeue();

                if ($node->left === null && $node->right === null) {
                    return $depth;
                }

                if ($node->left !== null) {
                    $queue->enqueue($node->left);
                }

                if ($node->right !== null) {
                    $queue->enqueue($node->right);
                }
            }

            $depth++;
        }

        return $depth;
    }
}

class TreeNode
{
    public ?int $val = null;

    public ?TreeNode $left = null;

    public ?TreeNode $right = null;

    public function __construct($val = 0, $left = null, $right = null)
    {
        $this->val = $val;
        $this->left = $left;
        $this->right = $right;
    }
}

File: Go/111. Minimum Depth of Binary Tree.go
package main

//goland:noinspection GoUnusedFunction
func minDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}

	depth := 0
	nodes := []*TreeNode{root}

	for len(nodes) > 0 {
		size := len(nodes)

		for i := 0; i < size; i++ {
			node := nodes[0]
			nodes = nodes[1:]

			if node.Left == nil && node.Right == nil {
				return depth + 1
			}

			if node.Left != nil {
				nodes = append(nodes, node.Left)
			}

			if node.Right != nil {
				nodes = append(nodes, node.Right)
			}
		}

		depth++
	}

	return depth
}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}
